File: crates/sdk/src/achievements.rs
//! Achievements (issue #34): read a user's own attestation history and
//! issue new attestations on this integrator's own behalf. Previously both
//! `Session` methods here returned `SdkError::NotImplemented`.
//!
//! **Reading** (`Session::achievements`) uses the identity's own bearer
//! token against `GET /me/achievements` — an identity reading its own full
//! history, the same "facts, not a per-consumer verdict" posture `GET
//! /attestations/{id}` (#33) already takes: `authenticity`/`validity` come
//! back as the server computed them. **Recognition is never present
//! here** — that's `avalon_protocol::achievements::recognize`, evaluated
//! by this integrator against its own `TrustRelationship`, matching ADR
//! #76's "recognition is contextual" rule. This SDK doesn't compute a
//! recognition verdict on the caller's behalf either; an integrator that wants
//! one filters [`VerifiedAttestation`]s through its own policy.
//!
//! **Issuing** (`Session::issue_achievement`) needs this integrator's own
//! signing key — the server never sees it, only a detached signature
//! (`AvalonConfig::integrator_slug`/`signing_key`). Two independent proofs go
//! out, mirroring every other issuer-credentialed endpoint in this repo
//! (`docs/architecture/issuers.md`): an ephemeral
//! challenge-response proving *this key* is making the HTTP call right now
//! (`POST /integrations/{slug}/challenge`), and a separate signature embedded in
//! the request body over the attestation's own canonical bytes, proving
//! *this key* specifically authorized *this* attestation — checked
//! independently server-side
//! (`avalon_chain::attestations::verify_authenticity`), not inferred from
//! the HTTP-level proof alone.
//!
//! Milestones (the App/Service equivalent, #324/#325) aren't wired up
//! here — this ticket's own scope is `achievements()`/`issue_achievement()`
//! specifically; a `milestones()`/`issue_milestone()` pair would follow the
//! same shape against `/integrations/{slug}/milestones/{key}/issue`.

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use ed25519_dalek::{Signer, SigningKey};
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use uuid::Uuid;

use crate::{SdkError, Session};

/// Mirrors `avalon_chain::attestations::Authenticity` at the wire level.
/// This crate defines its own copy rather than depending on `avalon-chain`
/// directly — an integrator has no business linking the verification
/// engine itself, only its result — the same posture every other SDK
/// response type in this crate already takes (e.g. `social::Friend`).
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum Authenticity {
    /// The embedded signature verifies against one of the issuer's keys.
    Authentic {
        /// Which of the issuer's keys verified it.
        key_id: String,
    },
    /// The embedded signature does not verify against any of the issuer's
    /// currently-known keys.
    NotAuthentic {
        /// Why — never used to make an authorization decision, only to
        /// explain a rejection to a developer.
        reason: String,
    },
}

/// Mirrors `avalon_protocol::achievements::Validity` at the wire level.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum Validity {
    /// Not revoked, and its definition isn't retired in a way that
    /// invalidates already-issued attestations.
    Valid,
    /// Revoked, or otherwise no longer honored.
    Invalid {
        /// Why — see `crate::attestations::Validity` server-side for the
        /// exact set of reasons.
        reason: String,
    },
}

/// One entry in an attestation's history (`"issued"`, plus `"revoked"` if
/// applicable, #85) — see `crate::attestations::AttestationHistoryEntry`
/// server-side.
#[derive(Debug, Clone, Deserialize)]
pub struct AttestationHistoryEntry {
    /// `"issued"` or `"revoked"`.
    pub event: String,
    /// When this history event happened.
    #[serde(with = "time::serde::rfc3339")]
    pub at: OffsetDateTime,
    /// A machine-readable reason code, present on a `"revoked"` entry.
    pub reason_code: Option<String>,
    /// A human-readable reason, present on a `"revoked"` entry.
    pub reason: Option<String>,
}

/// One attestation from the caller's own history, with its computed
/// authenticity and validity attached — deliberately no `recognition`
/// field, see module doc comment.
#[derive(Debug, Clone, Deserialize)]
pub struct VerifiedAttestation {
    /// This attestation's own id.
    pub id: Uuid,
    /// The issuer that issued it, e.g. `"game:<slug>"`.
    pub issuer: String,
    /// The identity this attestation is about.
    pub subject: Uuid,
    /// The achievement/milestone definition this attestation claims,
    /// e.g. `"game:<slug>:achievement:<key>"`.
    pub achievement: String,
    /// When it was issued.
    #[serde(with = "time::serde::rfc3339")]
    pub issued_at: OffsetDateTime,
    /// Whether the embedded signature actually verifies.
    pub authenticity: Authenticity,
    /// Whether it's still in force (not revoked).
    pub validity: Validity,
    /// Its full history — issuance, and revocation if any.
    pub history: Vec<AttestationHistoryEntry>,
}

#[derive(Deserialize)]
struct ChallengeResponse {
    challenge_id: Uuid,
    nonce: String,
}

#[derive(Serialize)]
struct IssueRequest {
    key_id: Uuid,
    signature: String,
}

#[derive(Deserialize)]
struct IssueResponse {
    id: Uuid,
}

/// The exact bytes this integrator's key signs to authorize an
/// attestation — must match
/// `avalon_protocol::achievements::attestation_signing_bytes` exactly.
/// This crate defines its own copy rather than depending on the server's
/// private construction: each side of the wire independently builds the
/// same canonical format, the same posture every other signed request in
/// this repo already takes (client and server never share a signing-bytes
/// function, only its documented shape).
fn attestation_signing_bytes(issuer_ref: &str, subject: Uuid, achievement: &str) -> Vec<u8> {
    format!("avalon:achievement.issued:v1:{issuer_ref}:{subject}:{achievement}").into_bytes()
}

/// The exact bytes this integrator's key signs to authorize a *bulk*
/// issuance (issue #495, implementing #492's decided shape) — must match
/// `avalon_protocol::achievements::bulk_attestation_signing_bytes` exactly.
/// Same "each side independently builds the same canonical format" posture
/// as [`attestation_signing_bytes`] above.
fn bulk_attestation_signing_bytes(
    issuer_ref: &str,
    subject: Uuid,
    achievements: &[String],
) -> Vec<u8> {
    let mut message =
        format!("avalon:achievement.issued.bulk:v1:{issuer_ref}:{subject}:").into_bytes();
    message.extend_from_slice(&(achievements.len() as u32).to_be_bytes());
    for achievement in achievements {
        message.extend_from_slice(&(achievement.len() as u32).to_be_bytes());
        message.extend_from_slice(achievement.as_bytes());
    }
    message
}

#[derive(Serialize)]
struct BulkClaimRequestWire {
    key: String,
}

#[derive(Serialize)]
struct BulkIssueRequest {
    key_id: Uuid,
    signature: String,
    claims: Vec<BulkClaimRequestWire>,
}

/// The attestation a successful bulk claim resulted in — a deliberately
/// smaller shape than the full server response (omitting `proof`, which
/// nothing in this SDK's own bulk caller needs), same minimalism
/// `Session::issue_achievement`'s bare `Uuid` return already takes for the
/// single-claim case.
#[derive(Debug, Clone, Deserialize)]
pub struct BulkIssuedAttestation {
    /// This attestation's own id.
    pub id: Uuid,
    /// The issuer that issued it, e.g. `"game:<slug>"`.
    pub issuer: String,
    /// The identity this attestation is about.
    pub subject: Uuid,
    /// The achievement definition this attestation claims, e.g.
    /// `"game:<slug>:achievement:<key>"`.
    pub achievement: String,
    /// When it was issued.
    #[serde(with = "time::serde::rfc3339")]
    pub issued_at: OffsetDateTime,
}

/// One claim's own outcome from a bulk issuance call — a bulk call is
/// never all-or-nothing (#495's own invariant): a claim referencing an
/// unknown or retired definition fails on its own, every other claim in
/// the same call still succeeds. Mirrors
/// `crates/server/src/achievements.rs::BulkClaimResult` at the wire level.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum BulkClaimOutcome {
    /// This claim was issued successfully.
    Issued {
        /// The achievement key this outcome is for.
        key: String,
        /// The resulting attestation.
        attestation: BulkIssuedAttestation,
    },
    /// This claim failed — every other claim in the same call may still
    /// have succeeded; check each [`BulkClaimOutcome`] independently.
    Failed {
        /// The achievement key this outcome is for.
        key: String,
        /// A stable, machine-readable reason code.
        code: String,
        /// A human-readable explanation of `code`.
        error: String,
    },
}

#[derive(Deserialize)]
struct BulkIssueResponseWire {
    results: Vec<BulkClaimOutcome>,
}

/// Mirrors `crate::attestations::ListMyAchievementsResponse` at the wire
/// level — issue #377 wrapped what was a bare array in a
/// `{ achievements, next_cursor }` envelope so `GET /me/achievements`
/// could be paginated. `fetch_achievements` below unwraps this and returns
/// just the first page's attestations, matching `Session::achievements`'s
/// existing, unpaginated public shape; a paginated/filtered entry point is
/// a follow-up, not built here (see issue #377's own PR for why the SDK
/// side was scoped out).
#[derive(Deserialize)]
struct ListMyAchievementsResponse {
    achievements: Vec<VerifiedAttestation>,
    #[allow(dead_code)]
    next_cursor: Option<Uuid>,
}

impl Session {
    pub(crate) async fn fetch_achievements(&self) -> Result<Vec<VerifiedAttestation>, SdkError> {
        // `limit=200` (the server's own max page size,
        // `attestations::MAX_ACHIEVEMENTS_PAGE_SIZE`) rather than the
        // default 50 — minimizes the behavior change from before #377's
        // pagination landed, though a caller with more than 200
        // attestations from a single identity now genuinely needs the
        // (not yet built) paginated entry point to see the rest.
        let response = crate::http::send(&self.http, &self.retry, true, |c| {
            c.get(format!("{}/me/achievements?limit=200", self.server_url))
                .bearer_auth(&self.token)
        })
        .await?;

        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }

        let page: ListMyAchievementsResponse = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        Ok(page.achievements)
    }

    /// Issue #47's own named example: this write carries a fresh
    /// `Idempotency-Key`, generated once per logical call and reused
    /// across every retry attempt of it (never regenerated per attempt —
    /// that would defeat the point), so a retried issuance replays the
    /// first attempt's result instead of minting a second attestation —
    /// see `crate::idempotency` server-side. The retry unit is the *whole*
    /// challenge-then-issue exchange, not just the final POST: a
    /// challenge is single-use, so retrying only the issue request with a
    /// possibly-already-consumed challenge id would fail differently
    /// rather than actually retry (`crate::http::retry_write`'s own doc
    /// comment).
    pub(crate) async fn submit_achievement_issuance(&self, key: &str) -> Result<Uuid, SdkError> {
        let idempotency_key = Uuid::new_v4().to_string();
        crate::http::retry_write(&self.retry, || {
            self.attempt_achievement_issuance(key, &idempotency_key)
        })
        .await
    }

    async fn attempt_achievement_issuance(
        &self,
        key: &str,
        idempotency_key: &str,
    ) -> Result<Uuid, SdkError> {
        let slug = self
            .integrator_slug
            .as_deref()
            .ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key_bytes = self.signing_key.ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key = SigningKey::from_bytes(&signing_key_bytes);
        let key_id: Uuid = self
            .integrator_key_id
            .parse()
            .map_err(|_| SdkError::MissingIssuerCredentials)?;

        // Proof one: this key is making this HTTP call, right now. Not
        // retried by `http::send` itself — a failed attempt here means
        // `retry_write` redoes this whole function, fetching a fresh
        // challenge along with it.
        let challenge_response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.post(format!(
                "{}/integrations/{}/challenge",
                self.server_url, slug
            ))
        })
        .await?;
        if !challenge_response.status().is_success() {
            return Err(crate::http::map_error_response(challenge_response).await);
        }
        let challenge: ChallengeResponse = challenge_response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        let nonce = BASE64
            .decode(&challenge.nonce)
            .map_err(|_| SdkError::MissingIssuerCredentials)?;
        let challenge_signature = signing_key.sign(&nonce);

        // Proof two: this key specifically authorized this attestation —
        // independent of the challenge-response above, checked
        // server-side against the same canonical bytes.
        let subject = self.identity.id.0;
        let issuer_ref = format!("game:{slug}");
        let achievement = format!("game:{slug}:achievement:{key}");
        let signing_bytes = attestation_signing_bytes(&issuer_ref, subject, &achievement);
        let signature = signing_key.sign(&signing_bytes);

        let response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.post(format!(
                "{}/integrations/{}/achievements/{}/issue",
                self.server_url, slug, key
            ))
            .header("x-avalon-integrator-key-id", &self.integrator_key_id)
            .header(
                "x-avalon-integrator-challenge-id",
                challenge.challenge_id.to_string(),
            )
            .header(
                "x-avalon-integrator-signature",
                BASE64.encode(challenge_signature.to_bytes()),
            )
            .header("x-avalon-identity-id", subject.to_string())
            .header("idempotency-key", idempotency_key)
            .json(&IssueRequest {
                key_id,
                signature: BASE64.encode(signature.to_bytes()),
            })
        })
        .await?;

        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }

        let body: IssueResponse = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        Ok(body.id)
    }

    /// Issue #495 (implementing #492's decided shape): the bulk-issuance
    /// counterpart to [`Session::submit_achievement_issuance`] — one
    /// challenge-response, one signature over the whole ordered `keys`
    /// list, every claim still becoming its own ordinary attestation
    /// server-side. Same idempotency/retry posture as the single-claim
    /// call: the *whole* challenge-then-bulk-issue exchange is the retry
    /// unit, never just the final POST.
    pub(crate) async fn submit_bulk_achievement_issuance(
        &self,
        keys: &[&str],
    ) -> Result<Vec<BulkClaimOutcome>, SdkError> {
        let idempotency_key = Uuid::new_v4().to_string();
        crate::http::retry_write(&self.retry, || {
            self.attempt_bulk_achievement_issuance(keys, &idempotency_key)
        })
        .await
    }

    async fn attempt_bulk_achievement_issuance(
        &self,
        keys: &[&str],
        idempotency_key: &str,
    ) -> Result<Vec<BulkClaimOutcome>, SdkError> {
        let slug = self
            .integrator_slug
            .as_deref()
            .ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key_bytes = self.signing_key.ok_or(SdkError::MissingIssuerCredentials)?;
        let signing_key = SigningKey::from_bytes(&signing_key_bytes);
        let key_id: Uuid = self
            .integrator_key_id
            .parse()
            .map_err(|_| SdkError::MissingIssuerCredentials)?;

        let challenge_response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.post(format!(
                "{}/integrations/{}/challenge",
                self.server_url, slug
            ))
        })
        .await?;
        if !challenge_response.status().is_success() {
            return Err(crate::http::map_error_response(challenge_response).await);
        }
        let challenge: ChallengeResponse = challenge_response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        let nonce = BASE64
            .decode(&challenge.nonce)
            .map_err(|_| SdkError::MissingIssuerCredentials)?;
        let challenge_signature = signing_key.sign(&nonce);

        let subject = self.identity.id.0;
        let issuer_ref = format!("game:{slug}");
        let achievements: Vec<String> = keys
            .iter()
            .map(|key| format!("game:{slug}:achievement:{key}"))
            .collect();
        let signing_bytes = bulk_attestation_signing_bytes(&issuer_ref, subject, &achievements);
        let signature = signing_key.sign(&signing_bytes);

        let response = crate::http::send(&self.http, &self.retry, false, |c| {
            c.post(format!(
                "{}/integrations/{}/achievements/bulk-issue",
                self.server_url, slug
            ))
            .header("x-avalon-integrator-key-id", &self.integrator_key_id)
            .header(
                "x-avalon-integrator-challenge-id",
                challenge.challenge_id.to_string(),
            )
            .header(
                "x-avalon-integrator-signature",
                BASE64.encode(challenge_signature.to_bytes()),
            )
            .header("x-avalon-identity-id", subject.to_string())
            .header("idempotency-key", idempotency_key)
            .json(&BulkIssueRequest {
                key_id,
                signature: BASE64.encode(signature.to_bytes()),
                claims: keys
                    .iter()
                    .map(|key| BulkClaimRequestWire {
                        key: key.to_string(),
                    })
                    .collect(),
            })
        })
        .await?;

        if !response.status().is_success() {
            return Err(crate::http::map_error_response(response).await);
        }

        let body: BulkIssueResponseWire = response
            .json()
            .await
            .map_err(|e| SdkError::Protocol(e.to_string()))?;
        Ok(body.results)
    }
}

File: bindings/csharp/AvalonSdk/Achievements.cs
// Achievements (issue #34, wired here for #51) — read a user's own attestation
// history and issue a new attestation on this integrator's own behalf.
// Mirrors crates/sdk/src/achievements.rs's single-claim (non-bulk,
// non-revocation) surface; bulk issuance (#495) and revocation (#498) are
// separate, later Rust-side tickets with no C# equivalent yet — #51's own
// invariant ("never exposes a method the Rust SDK doesn't have, and vice
// versa") is about keeping the two surfaces from silently diverging, not
// about building every Rust method on day one.
//
// Reading (GetAchievementsAsync) uses the identity's own bearer token
// against GET /me/achievements — authenticity/validity come back as the
// server computed them; recognition is deliberately never present (ADR #76:
// authenticity, validity, and recognition are three separate questions —
// this SDK never collapses them into one boolean).
//
// Issuing (IssueAchievementAsync) needs this integrator's own signing key —
// the server never sees it, only a detached signature. Two independent
// proofs go out: an ephemeral challenge-response proving this key is making
// the HTTP call right now, and a separate signature embedded in the request
// body over the attestation's own canonical bytes.

using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace Avalon.Sdk
{
    /// <summary>Mirrors <c>avalon_chain::attestations::Authenticity</c> at the wire level —
    /// whether the embedded signature actually verifies against one of the issuer's keys.
    /// A plain data class rather than a discriminated union: <see cref="Status"/> carries the
    /// tag ("authentic" / "not_authentic") and only the field that status defines is
    /// populated.</summary>
    public sealed class Authenticity
    {
        [JsonPropertyName("status")]
        public string Status { get; set; } = "";

        /// <summary>Which of the issuer's keys verified it — present only when
        /// <see cref="IsAuthentic"/>.</summary>
        [JsonPropertyName("key_id")]
        public string? KeyId { get; set; }

        /// <summary>Why it doesn't verify — present only when not authentic. Never used to make
        /// an authorization decision, only to explain a rejection to a developer.</summary>
        [JsonPropertyName("reason")]
        public string? Reason { get; set; }

        public bool IsAuthentic => Status == "authentic";
    }

    /// <summary>Mirrors <c>avalon_protocol::achievements::Validity</c> at the wire level —
    /// whether an attestation is still in force (not revoked).</summary>
    public sealed class Validity
    {
        [JsonPropertyName("status")]
        public string Status { get; set; } = "";

        /// <summary>Present only when not valid.</summary>
        [JsonPropertyName("reason")]
        public string? Reason { get; set; }

        public bool IsValid => Status == "valid";
    }

    /// <summary>One entry in an attestation's history ("issued", plus "revoked" if applicable).
    /// Mirrors <c>AttestationHistoryEntry</c> server-side.</summary>
    public sealed class AttestationHistoryEntry
    {
        [JsonPropertyName("event")]
        public string Event { get; set; } = "";

        [JsonPropertyName("at")]
        public DateTimeOffset At { get; set; }

        [JsonPropertyName("reason_code")]
        public string? ReasonCode { get; set; }

        [JsonPropertyName("reason")]
        public string? Reason { get; set; }
    }

    /// <summary>One attestation from the caller's own history, with its computed authenticity
    /// and validity attached — deliberately no combined "trusted" field; see this file's own
    /// header comment (ADR #76). Mirrors <c>avalon_sdk::achievements::VerifiedAttestation</c>.
    /// </summary>
    public sealed class VerifiedAttestation
    {
        [JsonPropertyName("id")]
        public Guid Id { get; set; }

        /// <summary>The issuer that issued it, e.g. "game:&lt;slug&gt;".</summary>
        [JsonPropertyName("issuer")]
        public string Issuer { get; set; } = "";

        [JsonPropertyName("subject")]
        public Guid Subject { get; set; }

        /// <summary>The achievement definition this attestation claims, e.g.
        /// "game:&lt;slug&gt;:achievement:&lt;key&gt;".</summary>
        [JsonPropertyName("achievement")]
        public string Achievement { get; set; } = "";

        [JsonPropertyName("issued_at")]
        public DateTimeOffset IssuedAt { get; set; }

        [JsonPropertyName("authenticity")]
        public Authenticity Authenticity { get; set; } = new Authenticity();

        [JsonPropertyName("validity")]
        public Validity Validity { get; set; } = new Validity();

        [JsonPropertyName("history")]
        public List<AttestationHistoryEntry> History { get; set; } = new List<AttestationHistoryEntry>();
    }

    // ListMyAchievementsResponse stays hand-written, keeping List<VerifiedAttestation>: the
    // generated Avalon.Sdk.Generated.ListMyAchievementsResponse's Achievements field is typed
    // as ICollection<AttestationReadResponse>, and AttestationReadResponse is one of the
    // deliberately-excluded oneOf schemas (its nested AuthenticityResponse/ValidityResponse
    // generate as empty stub classes) — migrating this DTO would silently drop every
    // authenticity/validity field this file exists to carry.
    internal sealed class ListMyAchievementsResponse
    {
        [JsonPropertyName("achievements")]
        public List<VerifiedAttestation> Achievements { get; set; } = new List<VerifiedAttestation>();

        [JsonPropertyName("next_cursor")]
        public Guid? NextCursor { get; set; }
    }

    public sealed partial class Session
    {
        /// <summary>Signs <paramref name="message"/> with this session's configured
        /// <see cref="SigningKey"/> (a 32-byte Ed25519 seed) — pure-managed Ed25519 via
        /// BouncyCastle, no native dependency, so this stays Unity/IL2CPP-safe.</summary>
        private byte[] SignWithIssuerKey(byte[] message)
        {
            var privateKey = new Ed25519PrivateKeyParameters(SigningKey, 0);
            var signer = new Ed25519Signer();
            signer.Init(true, privateKey);
            signer.BlockUpdate(message, 0, message.Length);
            return signer.GenerateSignature();
        }

        /// <summary>The exact bytes this integrator's key signs to authorize an attestation —
        /// must match <c>avalon_protocol::achievements::attestation_signing_bytes</c> exactly.
        /// This SDK defines its own copy rather than depending on the server's private
        /// construction, same posture as every other signed request in this repo.</summary>
        private static byte[] AttestationSigningBytes(string issuerRef, Guid subject, string achievement) =>
            Encoding.UTF8.GetBytes($"avalon:achievement.issued:v1:{issuerRef}:{subject}:{achievement}");

        /// <summary>POST /integrations/{slug}/challenge — an ephemeral, single-use
        /// challenge proving this integrator's key is making this HTTP call right now.</summary>
        private async Task<Avalon.Sdk.Generated.IntegratorChallengeResponse> RequestChallengeAsync(string slug, CancellationToken ct)
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{ServerUrl}/integrations/{slug}/challenge");
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            return await ReadJsonAsync<Avalon.Sdk.Generated.IntegratorChallengeResponse>(response, ct).ConfigureAwait(false);
        }

        /// <summary>GET /me/achievements — requires achievements.read. This identity's own
        /// attestation history; recognition is left to the caller's own policy (see this file's
        /// header comment). Mirrors the Rust SDK's <c>Session::achievements()</c>.</summary>
        public async Task<IReadOnlyList<VerifiedAttestation>> GetAchievementsAsync(CancellationToken ct = default)
        {
            Require("achievements.read");

            // limit=200 — the server's own max page size; a caller with more than 200
            // attestations needs a paginated entry point this SDK doesn't expose yet, same
            // documented gap as the Rust SDK's own `fetch_achievements`.
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{ServerUrl}/me/achievements?limit=200");
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", Token);
            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            var page = await ReadJsonAsync<ListMyAchievementsResponse>(response, ct).ConfigureAwait(false);
            return page.Achievements;
        }

        /// <summary>Issues <paramref name="key"/> (an achievement defined by this integrator) to
        /// this session's own identity, signed locally with <see cref="SigningKey"/>. Returns the
        /// new attestation's id. Throws <see cref="MissingIssuerCredentialsException"/>, without
        /// making any HTTP call, if <see cref="IntegratorSlug"/>/<see cref="SigningKey"/> weren't
        /// configured. Mirrors the Rust SDK's <c>Session::issue_achievement()</c>.</summary>
        public async Task<Guid> IssueAchievementAsync(string key, CancellationToken ct = default)
        {
            Require("achievements.issue");

            if (IntegratorSlug is null || SigningKey is null)
            {
                throw new MissingIssuerCredentialsException();
            }
            if (!Guid.TryParse(IntegratorKeyId, out var keyId))
            {
                throw new MissingIssuerCredentialsException();
            }

            var challenge = await RequestChallengeAsync(IntegratorSlug, ct).ConfigureAwait(false);
            var nonce = Convert.FromBase64String(challenge.Nonce);
            var challengeSignature = SignWithIssuerKey(nonce);

            var issuerRef = $"game:{IntegratorSlug}";
            var achievement = $"game:{IntegratorSlug}:achievement:{key}";
            var signingBytes = AttestationSigningBytes(issuerRef, IdentityGuid, achievement);
            var signature = SignWithIssuerKey(signingBytes);

            using var request = new HttpRequestMessage(
                HttpMethod.Post, $"{ServerUrl}/integrations/{IntegratorSlug}/achievements/{key}/issue");
            request.Headers.Add("x-avalon-integrator-key-id", IntegratorKeyId);
            request.Headers.Add("x-avalon-integrator-challenge-id", challenge.ChallengeId.ToString());
            request.Headers.Add("x-avalon-integrator-signature", Convert.ToBase64String(challengeSignature));
            request.Headers.Add("x-avalon-identity-id", IdentityGuid.ToString());
            // A fresh Idempotency-Key per call (issue #47): a caller that retries this whole
            // call after a dropped response mints a new key, same as the Rust SDK's
            // `submit_achievement_issuance` — this method doesn't itself retry.
            request.Headers.Add("idempotency-key", Guid.NewGuid().ToString());
            request.Content = new StringContent(
                JsonSerializer.Serialize(new Avalon.Sdk.Generated.IssueAttestationRequest
                {
                    KeyId = keyId,
                    Signature = Convert.ToBase64String(signature),
                }),
                Encoding.UTF8, "application/json");

            using var response = await Http.SendAsync(request, ct).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode)
            {
                throw ServerError(response.StatusCode);
            }
            var body = await ReadJsonAsync<Avalon.Sdk.Generated.AttestationResponse>(response, ct).ConfigureAwait(false);
            return body.Id;
        }
    }
}
